### PR TITLE
Adds new action timing options

### LINKED ISSUE #711: Sort out (outdated) action timings
Actions can currently be executed at 5 different occasions (see [wiki](https://github.com/precice/precice/wiki/Actions-Configuration#basics-and-pre-implemented-actions)):

* `regular-prior`: 
* `regular-post`: 
* `on-exchange-prior`: 
* `on-exchange-post`: 
* `on-time-window-complete-post`: 

Their meaning is roughly:
* `prior` and `post` means before or after communication.
* The difference between `regular` and `on-exchange` is whether data is exchanged or not. So, meaning whether the end of the time window is already reached or not. 
* I currently don't understand why `on-time-window-complete-post` is needed.

A further problem is that we don't distinguish whether the action is executed before or after a potential corresponding mapping. Actually, there is even a crucial difference here (looks like a bug):

In **initialize**:
https://github.com/precice/precice/blob/04bef9d14afdc94229f6788b2e29be5d465f9abd/src/precice/impl/SolverInterfaceImpl.cpp#L257-L264

In **advance**:
https://github.com/precice/precice/blob/04bef9d14afdc94229f6788b2e29be5d465f9abd/src/precice/impl/SolverInterfaceImpl.cpp#L368-L379

Purpose of this issue is to properly discuss and define which timings we actually need and to then implement them. Depending on the outcome of the discussion, we need to be careful about backwards compatibility, though I believe that actions are not used a lot currently.

### PR DESCRIPTION
**Pull request by KyleDavisSA**
This PR adds new action timing options:

- WRITE_MAPPING_PRIOR,          everytime, before write mapping 
- WRITE_MAPPING_POST,            everytime, after write mapping and before advancing cpl scheme
- READ_MAPPING_PRIOR,            everytime, after advancing cpl scheme and before read mapping 
- READ_MAPPING_POST,               everytime, after read mapping 

Relation to issue https://github.com/precice/precice/issues/711

Resolves #711

### CODE REVIEW

File: src/action/config/ActionConfiguration.cpp [hunk 1]
```diff
@@ -304,15 +309,27 @@ action::Action::Timing ActionConfiguration::getTiming() const
   PRECICE_TRACE(_configuredAction.timing);
   action::Action::Timing timing;
   if (_configuredAction.timing == VALUE_REGULAR_PRIOR) {
-    timing = action::Action::ALWAYS_PRIOR;
+    timing = action::Action::WRITE_MAPPING_PRIOR;
+    PRECICE_WARN("Regular-prior action timings will no longer be supported. Regular-prior will now revert to Write-mapping-prior which performs "
+                  "the action timings before writeMappingData and before the coupling update.");
```
Comment (MakisH) [inline src/action/config/ActionConfiguration.cpp:314]: For consistency with the documentation, more user-friendly terms:
```suggestion
    PRECICE_WARN("Regular-prior action timings will no longer be supported. Regular-prior will now revert to write-mapping-prior which performs "
                  "the action before a write mapping and before the coupling update.");
```

File: src/action/config/ActionConfiguration.cpp [hunk 2]
```diff
@@ -304,15 +309,27 @@ action::Action::Timing ActionConfiguration::getTiming() const
...
   } else if (_configuredAction.timing == VALUE_REGULAR_POST) {
-    timing = action::Action::ALWAYS_POST;
+    timing = action::Action::READ_MAPPING_PRIOR;
+    PRECICE_WARN("Regular-post action timings will no longer be supported. Regular-post will now revert to Read-mapping-prior which performs "
+                  "the action timings before ReadMappingData but after the coupling update.");
```
Comment (MakisH) [inline src/action/config/ActionConfiguration.cpp:318]:
```suggestion
    PRECICE_WARN("Regular-post action timings will no longer be supported. Regular-post will now revert to read-mapping-prior which performs "
                  "the action after the coupling update and before a read mapping.");
```

File: src/precice/impl/SolverInterfaceImpl.cpp [hunk 3]
```diff
@@ -81,19 +82,12 @@ SolverInterfaceImpl::SolverInterfaceImpl(
...
 {
-  PRECICE_CHECK(!_accessorName.empty(), "This participant's name is an empty string. When constructing a preCICE interface "
-                                        "you need to pass the name of the participant as first argument to the constructor.");
-  PRECICE_CHECK(_accessorProcessRank >= 0,
-                "The solver process index needs to be a non-negative number, not: "
-                    << _accessorProcessRank << ". Please check the value given when constructing a preCICE interface.");
-  PRECICE_CHECK(_accessorCommunicatorSize >= 1,
-                "The solver process size needs to be a positive number, not: "
-                    << _accessorCommunicatorSize << ". Please check the value given when constructing a preCICE interface.");
+  PRECICE_CHECK(!_accessorName.empty(), "Accessor has to be named!");
```
Comment (MakisH) [inline src/precice/impl/SolverInterfaceImpl.cpp:85]: These error messages seem to be worsening (reverted changes). Maybe merge from develop?

Comment (MakisH) [reply src/precice/impl/SolverInterfaceImpl.cpp:85]: Check also the rest of the diff.

File: src/precice/impl/SolverInterfaceImpl.cpp [hunk 4]
```diff
@@ -319,13 +312,6 @@ double SolverInterfaceImpl::initialize()
 void SolverInterfaceImpl::initializeData()
 {
   PRECICE_TRACE();
-  PRECICE_CHECK(!_hasInitializedData, "initializeData() may only be called once.");
-  PRECICE_CHECK(_state != State::Finalized, "initializeData() cannot be called after finalize().")
-  PRECICE_CHECK(_state == State::Initialized, "initialize() has to be called before initializeData()");
-  PRECICE_ASSERT(_couplingScheme->isInitialized());
-  PRECICE_CHECK(not(_couplingScheme->sendsInitializedData() && isActionRequired(constants::actionWriteInitialData())),
-                "Initial data has to be written to preCICE by calling an appropriate write...Data() function before calling initializeData(). "
-                "Did you forget to call markActionFulfilled(precice::constants::actionWriteInitialData()) after writing initial data?");
```
Comment (MakisH) [inline src/precice/impl/SolverInterfaceImpl.cpp:328]: Why are these removed?

File: src/precice/impl/SolverInterfaceImpl.cpp [hunk 5]
```diff
@@ -377,12 +362,8 @@ double SolverInterfaceImpl::advance(
   Event                    e("advance", precice::syncMode);
   utils::ScopedEventPrefix sep("advance/");
 
-  PRECICE_CHECK(_state != State::Constructed, "initialize() has to be called before advance().");
-  PRECICE_CHECK(_state != State::Finalized, "advance() cannot be called after finalize().")
-  PRECICE_ASSERT(_couplingScheme->isInitialized());
-  PRECICE_CHECK(isCouplingOngoing(), "advance() cannot be called when isCouplingOngoing() returns false.");
-  PRECICE_CHECK((not _couplingScheme->receivesInitializedData() && not _couplingScheme->sendsInitializedData()) || (_hasInitializedData),
-                "initializeData() needs to be called before advance if data has to be initialized.");
```
Comment (MakisH) [inline src/precice/impl/SolverInterfaceImpl.cpp:385]: Why these changes here?

File: src/precice/impl/SolverInterfaceImpl.cpp [hunk 6]
```diff
@@ -538,51 +523,39 @@ int SolverInterfaceImpl::getDimensions() const
 bool SolverInterfaceImpl::isCouplingOngoing() const
 {
   PRECICE_TRACE();
-  PRECICE_CHECK(_state != State::Constructed, "initialize() has to be called before isCouplingOngoing() can be evaluated.");
-  PRECICE_CHECK(_state != State::Finalized, "isCouplingOngoing() cannot be called after finalize().");
```
Comment (MakisH) [inline src/precice/impl/SolverInterfaceImpl.cpp:542]: Many checks are being removed. Is this up-to-date with develop?

**Comment by MakisH** [comment]
- [ ] We also need to update the [wiki page](https://github.com/precice/precice/wiki/Actions-Configuration)

**Comment by KyleDavisSA** [comment]
There was an error along the way an this PR was not uptodate with develop, resulting in changes to the `SOlverInterfaceImpl` that was not intended

**Comment by fsimonis** [comment]
I reverted the changes.
Please let me merge this PR later. I'll make sure to remove the changes altogether.

**Comment by uekerman** [comment]
Don't forget to call 
```
performDataActions(timings, time, computedTimestepLength, timeWindowComputedPart, timeWindowSize);
```
each time you add actions. Otherwise, nothing well happen.

**Comment by fsimonis** [comment]
@uekerman So all these calls to insert should really be direct calls to performActions.
After all, we have precise timings now. Is that correct?

**Comment by fsimonis** [comment]
Just talked to @uekerman about the state of the actions. There are currently a few problems the way we handle them.

I'll take this PR over and add a few additional changes.

**Comment by fsimonis** [comment]
So, I changed the timings to be more explicit. So more sets to track.

File: src/action/config/ActionConfiguration.cpp [hunk 7]
```diff
@@ -304,15 +320,27 @@ action::Action::Timing ActionConfiguration::getTiming() const
   PRECICE_TRACE(_configuredAction.timing);
   action::Action::Timing timing;
   if (_configuredAction.timing == VALUE_REGULAR_PRIOR) {
-    timing = action::Action::ALWAYS_PRIOR;
+    timing = action::Action::WRITE_MAPPING_PRIOR;
+    PRECICE_WARN("Regular-prior action timings will no longer be supported. Regular-prior will now revert to write-mapping-prior which performs "
```
Comment (uekerman) [inline src/action/config/ActionConfiguration.cpp:324]:
```suggestion
    PRECICE_WARN("Regular-prior action timing is deprecated. Regular-prior will now revert to write-mapping-prior which performs "
```

File: src/action/config/ActionConfiguration.cpp [hunk 8]
```diff
@@ -304,15 +320,27 @@ action::Action::Timing ActionConfiguration::getTiming() const
...
+                 "the action before a write mapping and before the coupling update.");
   } else if (_configuredAction.timing == VALUE_REGULAR_POST) {
-    timing = action::Action::ALWAYS_POST;
+    timing = action::Action::READ_MAPPING_PRIOR;
+    PRECICE_WARN("Regular-post action timings will no longer be supported. Regular-post will now revert to read-mapping-prior which performs "
```
Comment (uekerman) [inline src/action/config/ActionConfiguration.cpp:328]:
```suggestion
    PRECICE_WARN("Regular-post action timings is deprecated. Regular-post will now revert to read-mapping-prior which performs "
```

File: src/precice/impl/SolverInterfaceImpl.cpp [hunk 9]
```diff
@@ -407,31 +406,26 @@ double SolverInterfaceImpl::advance(
...
-  timings.clear();
-  timings.insert(action::Action::ALWAYS_POST);
   if (_couplingScheme->hasDataBeenReceived()) {
-    timings.insert(action::Action::ON_EXCHANGE_POST);
+    performDataActions({action::Action::ON_EXCHANGE_POST}, time, computedTimestepLength, timeWindowComputedPart, timeWindowSize);
   }
   if (_couplingScheme->isTimeWindowComplete()) {
-    timings.insert(action::Action::ON_TIME_WINDOW_COMPLETE_POST);
+    performDataActions({action::Action::ON_TIME_WINDOW_COMPLETE_POST}, time, computedTimestepLength, timeWindowComputedPart, timeWindowSize);
   }
```
Comment (uekerman) [inline src/precice/impl/SolverInterfaceImpl.cpp:423]: Shouldn't we move that behind `mapReadData`?

File: src/precice/tests/recorder-action-explicit.xml [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+    <participant name="SolverOne">
+      <use-mesh name="MeshOne" provide="yes"/>
+      <write-data name="Forces"     mesh="MeshOne" />
+      <read-data  name="Velocities" mesh="MeshOne" />
+      <action:recorder timing="on-time-window-complete-post" mesh="MeshOne">
+      <action:recorder timing="write-mapping-prior" mesh="MeshOne">
+      <action:recorder timing="write-mapping-post" mesh="MeshOne">
+      <action:recorder timing="read-mapping-prior" mesh="MeshOne">
+      <action:recorder timing="read-mapping-post" mesh="MeshOne">
```
Comment (uekerman) [inline src/precice/tests/recorder-action-explicit.xml:32]: The more interesting use case will probably be the one where we have this at the participant where also the mapping happens. This one could be a valid check though as well.

**Comment by fsimonis** [comment]
What are we doing with the unconditional read mapping in initialize data?

**Comment by uekerman** [comment]
I would say let's carry this to a new issue and wrap up here. It's not completely connected to this one here.
